index.js

```diff
@@ -97,121 +168,195 @@ Query.scale = function(_domain) {
       return result(false,true);
     };
     return res;   
-  };
+
+  });
 };
 
 Query.$valueFn = function(d) { return d; };
 
-// Pick returns a filtered subset of the records
-Query.where = function(_filters, $valueFn) {
-  var self = this;
-  return function $property(ref) {
-  
-    // Provide pipe delimited filtering
-    ref = ref.split(reSplitter).sort();
-    if (ref.length > 1)
-      // Solve for the first one, and then the remainder
-      return [ref[0],function(q) {
-        return [{q:q},'q.where.'+ref.slice(1).join(reSplitter),Object];
-      }];
+function generateEvaluateConditionFn(self, ast, $global, _filters, $valueFn, pipeOperation) {
+  if (ast.piped) {
+    let operations = ast.piped.map(a => generateEvaluateConditionFn(self, a, $global, _filters, $valueFn));
+    return item => Promise.map(operations, operation => operation(item))
+              .then(results => {
+                if (pipeOperation === 'and') {
+                  return !results.some(inverseIdentity);
+                }
+                else {
+                  return results.some(identity);
+                }
+              });
+  }
+  else if (ast.not) {
+    let fn = generateEvaluateConditionFn(self, ast.not, $global, _filters, $valueFn, 'or');
+    return item => fn(item).then(inverseIdentity);
+  }
+  else if (ast.or) {
+    return generateEvaluateConditionFn(self, ast.or, $global, _filters, $valueFn, 'or');
+  }
+  else if (ast.and) {
+    return generateEvaluateConditionFn(self, ast.and, $global, _filters, $valueFn, 'and');
+  }
+  else if (ast.equation) {
+    let path = astToCluesPath(ast.equation.left);
+    let target = ast.equation.right;
+    if (target === 'true') target = true;
+    if (target === 'false') target = false;
     
-    ref = ref[0];
+    return async item => {
+      let leftSide = clues(item, path, $global).catch(noop);
+      let rightSide = target;
+      
+      if (target.quoted) {
+        target = target.quoted;
+      }
+      if (target.remoteLink) {
+        try {
+          rightSide = await clues({q:self}, 'q.'+astToCluesPath(target.remoteLink), $global, 'item');
```

Comment:
if i remove `'item' as a ref here, you can't access global.  I'm not sure if that's good or not.  i don't think we have secrets directly accessible in global --- and there is no way to get the results of this really.   

maybe you can brute force it?  `data.a<database_password` and then you binary search it?

Reply:
Yeah ideally nothing with possible user input should be called with a "caller" implying an internal function